crates/librqbit/src/blocklist.rs

```diff
@@ -0,0 +1,339 @@
+use anyhow::Result;
+use async_compression::tokio::bufread::GzipDecoder;
+use futures::TryStreamExt;
+use intervaltree::IntervalTree;
+use std::net::IpAddr;
+use std::pin::Pin;
+use std::str::FromStr;
+use tokio::io::AsyncRead;
+use tokio::{io::AsyncBufReadExt, io::BufReader};
+use tokio_util::io::StreamReader;
+use tracing::{debug, info, trace};
+
+pub struct Blocklist {
+    // Separate trees for IPv4 and IPv6 since they have different numeric ranges
+    ipv4_ranges: IntervalTree<u32, ()>,
+    ipv6_ranges: IntervalTree<u128, ()>,
+}
+
+impl Blocklist {
+    pub fn empty() -> Self {
+        return Self::new(
+            &Vec::<std::ops::Range<u32>>::new(),
+            &Vec::<std::ops::Range<u128>>::new(),
+        );
+    }
+
+    pub fn new(
+        ipv4_ranges: &Vec<std::ops::Range<u32>>,
```

Comment:
this can be simplified like this:

```
pub fn new(
    ipv4_ranges: impl IntoIterator<Item=std::ops::Range<Ipv4Addr>>,
    ...
) -> Self {
    ipv4_ranges: IntervalTree::from_iter(ipv4_ranges.map(|r| (r, ())),
    ...
```